### PR TITLE
[BEAM-1820] Make Source.getDefaultOutputCoder() @Nullable

### PR DESCRIPTION
**Pull request by lgajowy**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by lgajowy** [comment]
Hi, @jkff! Could you take a look at this contribution? I'm not completely sure if that's all. If not, could you comment and maybe answer my questions [here](https://issues.apache.org/jira/browse/BEAM-1820?focusedCommentId=16078407&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16078407)?

**Comment by lgajowy** [comment]
@chamikaramj FYI

**Comment by jkff** [comment]
retest this please

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java [hunk 1]
```diff
@@ -76,6 +76,7 @@
   private static final Logger LOG = LoggerFactory.getLogger(UnboundedReadFromBoundedSource.class);
 
   private final BoundedSource<T> source;
+  private Coder<T> defaultOutputCoder;
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: Transform classes can not have mutable state. Moreover, having getDefaultOutputCoder() return different values before and after expand() was called seems very brittle - the SDK or runners make no guarantees about the order in which methods will be called on a PTransform (instead, we require that PTransform's are immutable, so the concept of call order is irrelevant)

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java [hunk 2]
```diff
@@ -86,13 +87,17 @@ public UnboundedReadFromBoundedSource(BoundedSource<T> source) {
...
+
+    defaultOutputCoder = collection.getCoder();
+
+    return collection;
   }
 
   @Override
   protected Coder<T> getDefaultOutputCoder() {
-    return source.getDefaultOutputCoder();
+    return defaultOutputCoder;
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: What if you just don't override this method? Same in other places in this PR.

It seems like getDefaultOutputCoder(), ultimately, is used to provide a Coder for a PCollection produced by a PTransform in case the PTransform didn't already .setCoder() on the PCollection explicitly - see https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollection.java#L137 - which means that, for any particular transform, you should be able to remove an override of getDefaultOutputCoder() and instead either call .setCoder() on the PCollection it produces, or even just rely on the other coder inference mechanisms (see code below highlight in the link above).

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java (hunk 1 above)
Comment (lgajowy) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: Ok, I'll apply this suggestion in the next commit.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java (hunk 2 above)
Comment (lgajowy) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: I didn't think of that this way, but after reading the code you mentioned it seems a far better idea. There are more PTransforms that need be modified the way you described. I'll also fix them in the next commit.

**Comment by lgajowy** [comment]
@jkff Thanks for the previous review! I updated the code. Could you take a look?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java [hunk 3]
```diff
@@ -86,16 +86,11 @@ public UnboundedReadFromBoundedSource(BoundedSource<T> source) {
 
   @Override
   public PCollection<T> expand(PBegin input) {
-    return input.getPipeline().apply(
+      return input.getPipeline().apply(
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: Remove unrelated indentation change

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java [hunk 4]
```diff
@@ -20,12 +20,14 @@
...
 import java.io.Serializable;
 import java.util.NoSuchElementException;
+import javax.annotation.Nullable;
 import org.apache.beam.sdk.annotations.Experimental;
 import org.apache.beam.sdk.coders.Coder;
 import org.apache.beam.sdk.transforms.display.DisplayData;
 import org.apache.beam.sdk.transforms.display.HasDisplayData;
 import org.joda.time.Instant;
 
+
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java]: Remove unrelated whitespace change

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java [hunk 5]
```diff
@@ -64,6 +66,7 @@
   /**
    * Returns the default {@code Coder} to use for the data read from this source.
    */
+  @Nullable
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java]: Now that I think of it, it would be better to have consistency between PTransform.getDefaultOutputCoder (throws CannotProvideCoderException if it can't provide a coder) and Source.getDefaultOutputCoder - the latter should probably also throw a CannotProvideCoderException, rather than be nullable.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java [hunk 6]
```diff
@@ -86,16 +86,11 @@ public UnboundedReadFromBoundedSource(BoundedSource<T> source) {
...
         Read.from(new BoundedToUnboundedSourceAdapter<>(source)));
   }
 
   @Override
-  protected Coder<T> getDefaultOutputCoder() {
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: Why is it safe to simply remove this override here without any replacement?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 7]
```diff
@@ -1192,12 +1186,8 @@ private ReadWithIds(UnboundedSource<T, ?> source) {
...
       public final PCollection<ValueWithRecordId<T>> expand(PInput input) {
         return PCollection.<ValueWithRecordId<T>>createPrimitiveOutputInternal(
-            input.getPipeline(), WindowingStrategy.globalDefault(), IsBounded.UNBOUNDED);
-      }
-
-      @Override
-      protected Coder<ValueWithRecordId<T>> getDefaultOutputCoder() {
-        return ValueWithRecordId.ValueWithRecordIdCoder.of(source.getDefaultOutputCoder());
+            input.getPipeline(), WindowingStrategy.globalDefault(), IsBounded.UNBOUNDED)
+            .setCoder(ValueWithRecordId.ValueWithRecordIdCoder.of(source.getDefaultOutputCoder()));
```
Comment (jkff) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: source.getDefaultOutputCoder() now can be null, right? I don't think you can pass a null coder to ValueWithRecordIdCoder.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java [hunk 8]
```diff
@@ -1275,15 +1265,11 @@ public StreamingBoundedRead(DataflowRunner runner, Read.Bounded<T> transform) {
     }
 
     @Override
-    protected Coder<T> getDefaultOutputCoder() {
```
Comment (jkff) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: Same question

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java [hunk 9]
```diff
@@ -114,12 +114,8 @@
             }
           }));
     }
-    return read.apply("StripIds", ParDo.of(new ValueWithRecordId.StripIdsDoFn<T>()));
-  }
 
-  @Override
-  protected Coder<T> getDefaultOutputCoder() {
-    return source.getDefaultOutputCoder();
+    return read.apply("StripIds", ParDo.of(new ValueWithRecordId.StripIdsDoFn<T>()));
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedReadFromUnboundedSource.java]: And here too

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java [hunk 10]
```diff
@@ -95,17 +94,12 @@ private Bounded(@Nullable String name, BoundedSource<T> source) {
...
-    }
-
-    @Override
     public final PCollection<T> expand(PBegin input) {
       source.validate();
 
       return PCollection.<T>createPrimitiveOutputInternal(input.getPipeline(),
           WindowingStrategy.globalDefault(), IsBounded.BOUNDED)
-          .setCoder(getDefaultOutputCoder());
+          .setCoder(source.getDefaultOutputCoder());
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: PCollection.setCoder() does not allow passing a null argument, but source.getDefaultOutputCoder() is nullable.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java [hunk 11]
```diff
@@ -163,16 +157,12 @@ private Unbounded(@Nullable String name, UnboundedSource<T, ?> source) {
...
-    }
-
-    @Override
     public final PCollection<T> expand(PBegin input) {
       source.validate();
 
       return PCollection.<T>createPrimitiveOutputInternal(
-          input.getPipeline(), WindowingStrategy.globalDefault(), IsBounded.UNBOUNDED);
+          input.getPipeline(), WindowingStrategy.globalDefault(), IsBounded.UNBOUNDED)
+          .setCoder(source.getDefaultOutputCoder());
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: Same here

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java (hunk 5 above)
Comment (lgajowy) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java]: You've got a point about the consistency. In that case I see it like this:

Source.java
```
public Coder<T> getDefaultOutputCoder() throws CannotProvideCoderException {
    throw new CannotProvideCoderException("Source.getDefaultOutputCoder called.");
  }
```

As the exception is checked, I have the following questions:

1. Should I throw further (throws clause) the exception in every source that extends the Source class  (CompressedSource, MicrobatchSource, etc)? Because it seems that there is no way to infer/know what Coder should it be. If not, could you give me an example of how should I handle it?

2. Is it ok that it's a breaking change to the API? There are some classes that are not marked as experimental. eg. Read.Bounded, like below: 

Read.java
```
    @Override
    protected Coder<T> getDefaultOutputCoder() throws CannotProvideCoderException {
      return source.getDefaultOutputCoder(); 
    }

    @Override
    public final PCollection<T> expand(PBegin input) throws CannotProvideCoderException {
      source.validate();

      return PCollection.<T>createPrimitiveOutputInternal(input.getPipeline(),
          WindowingStrategy.globalDefault(), IsBounded.BOUNDED)
          .setCoder(getDefaultOutputCoder());
    }
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java (hunk 4 above)
Comment (lgajowy) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java]: ok, I'll be more careful

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java (hunk 3 above)
Comment (lgajowy) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: ok

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java (hunk 5 above)
Comment (lgajowy) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java]: @jkff ping :)

Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Source.java]: 1. Yes, I think you should just add the `throws` clause in those cases.
2. You're right, this is a breaking change. However, the Source class itself is marked `@Experimental`. Arguably it shouldn't be, but... it is. In general, still, the threshold for changing the Source API should be very high, but this particular change would affect only users who are _calling_ Source.getDefaultOutputCoder (it would not affect users who are merely _implementing_ it - adding a throws to a base class method signature is a compatible change for subclasses), which I expect is an extremely small subset of advanced users. I think the benefit of this change outweighs this cost.

**Comment by lgajowy** [comment]
@jkff Here's another attempt. I applied your suggestion to throw CannotProvideCoderException. 

The basic idea is to throw it further in places where we cannot handle the exception. However, there are places where doing so would imply the need of declaring the exception not only in Source class but in other abstract classes too (see the extend() method in Read.java, SparkUnboundedSource.java or UnboundedReadFromBoundedSource.java classes). I found it messy because it would spread the exception everywhere. Besides that, in StateSpecFunctions.java we cannot add the throws clause because the class is implementing external API which we cannot modify. 

In those places I decided to rethrow the exception as RuntimeException as I've found no better way of dealing with it in such situations. It's far more better from passing null coder because it provides a meaningful exception message in case of no default coder.  Let me know if you agree with this solution.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java (hunk 8 above)
Comment (lgajowy) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java]: This was fixed in next commits.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java [hunk 12]
```diff
@@ -166,14 +167,18 @@ public Reader createReader(PipelineOptions options, Checkpoint<T> checkpoint)
...
-    public Coder<T> getDefaultOutputCoder() {
+    public Coder<T> getDefaultOutputCoder() throws CannotProvideCoderException {
       return boundedSource.getDefaultOutputCoder();
     }
 
     @SuppressWarnings({"rawtypes", "unchecked"})
     @Override
     public Coder<Checkpoint<T>> getCheckpointMarkCoder() {
-      return new CheckpointCoder<>(boundedSource.getDefaultOutputCoder());
+      try {
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: Hm, I guess getCheckpointMarkCoder should also throw this exception. But fine to leave as-is for this PR.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java [hunk 13]
```diff
@@ -95,17 +96,20 @@ private Bounded(@Nullable String name, BoundedSource<T> source) {
...
       source.validate();
-
-      return PCollection.<T>createPrimitiveOutputInternal(input.getPipeline(),
-          WindowingStrategy.globalDefault(), IsBounded.BOUNDED)
-          .setCoder(getDefaultOutputCoder());
+      try {
+        return PCollection.<T>createPrimitiveOutputInternal(input.getPipeline(),
+            WindowingStrategy.globalDefault(), IsBounded.BOUNDED)
+            .setCoder(getDefaultOutputCoder());
+      } catch (CannotProvideCoderException e) {
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: I don't think we should rethrow the exception here. It only means that the source wasn't able to provide a coder for us - so we should just not call .setCoder() ourselves, and the user will need to explicitly .setCoder() on the returned PCollection.

Actually I'm not sure why there even was a .setCoder() call in the original code - it is not present in the Read.Unbounded.expand() method. I think we should just mimic that method and remove this call.

Comment (lgajowy) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: Ok - I'll fix this in another commit.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java (hunk 12 above)
Comment (lgajowy) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: It looks as an easy and analogous fix - I'll post a proposal in a separate commit for this too (in case I missed something it will be easier to revert).

**Comment by lgajowy** [comment]
@jkff Thanks for the previous review. I applied your latest suggestions (two commits) - could you take a look?
The change in UnboundedSource.getCheckpointMarkCoder() seemed analogous to the one in Source.getDefaultOutputCoder() so I commited it as well. This way i find the whole issue more complete. Please let me know if you think different.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java [hunk 14]
```diff
@@ -104,8 +105,7 @@ private Bounded(@Nullable String name, BoundedSource<T> source) {
       source.validate();
 
       return PCollection.<T>createPrimitiveOutputInternal(input.getPipeline(),
-          WindowingStrategy.globalDefault(), IsBounded.BOUNDED)
-          .setCoder(getDefaultOutputCoder());
+          WindowingStrategy.globalDefault(), IsBounded.BOUNDED);
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: Could you also add a pair of unit tests to test this - where we successfully read (using a TestPipeline) from a bounded/unbounded source that does not override getDefaultOutputCoder, and the returned type of the source does not have a default coder, and instead the user (well, the test) calls .setCoder() on the resulting PCollection explicitly?

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java [hunk 15]
```diff
@@ -166,13 +167,13 @@ public Reader createReader(PipelineOptions options, Checkpoint<T> checkpoint)
...
     @Override
-    public Coder<T> getDefaultOutputCoder() {
+    public Coder<T> getDefaultOutputCoder() throws CannotProvideCoderException {
       return boundedSource.getDefaultOutputCoder();
     }
 
     @SuppressWarnings({"rawtypes", "unchecked"})
     @Override
-    public Coder<Checkpoint<T>> getCheckpointMarkCoder() {
+    public Coder<Checkpoint<T>> getCheckpointMarkCoder() throws CannotProvideCoderException {
```
Comment (jkff) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/UnboundedReadFromBoundedSource.java]: Hmm sorry, I just thought about this some more and realized that this is very problematic. Since now it is legal to have a source that does not provide a coder for itself - such a source can be used here too (in BoundedToUnboundedSourceAdapter). It looks like, in that case, getCheckpointMarkCoder() will fail - and this is a problem: when you just apply Read.from() to a source that doesn't provide a coder, you can simply set the coder explicitly on the resulting PCollection - but there's no such alternative for getCheckpointMarkCoder. And I don't think there should be - checkpoint mark coder of a source is an implementation detail of the source, so it's the source's responsibility to know this coder and the method probably should not be allowed to throw a CannotProvideCoderException.

What's worse - this class (UnboundedReadFromBoundedSource) is hidden from the user, and gets transparently applied by Dataflow runner at https://github.com/apache/beam/blob/master/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java#L365. That class (StreamingUnboundedRead) will call source.getDefaultOutputCoder() and will fail. Instead, it should infer the checkpoint mark coder for the BoundedToUnboundedSourceAdapter from the coder of the output PCollection - for that, StreamingUnboundedRead must stop being a ReflectiveRootOverrideFactory and should be an actual PTransformOverrideFactory and get the output PCollection from the AppliedPTransform.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/io/SparkUnboundedSource.java [hunk 16]
```diff
@@ -114,10 +115,16 @@
...
     // output the actual (deserialized) stream.
-    WindowedValue.FullWindowedValueCoder<T> coder =
-        WindowedValue.FullWindowedValueCoder.of(
-            source.getDefaultOutputCoder(),
-            GlobalWindow.Coder.INSTANCE);
+    WindowedValue.FullWindowedValueCoder<T> coder;
+
+    try {
+      coder = WindowedValue.FullWindowedValueCoder.of(
+          source.getDefaultOutputCoder(),
```
Comment (jkff) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/io/SparkUnboundedSource.java]: Here too - getDefaultOutputCoder() is allowed to fail and it simply means "the actual coder should be set by the user on the output PCollection" - i.e. if it failed but the user set the coder on the collection, this class should succeed.

In other words, we can not rethrow the exception here, and more like - we must not call source.getDefaultOutputCoder. More like: source.getDefaultOutputCoder() must not be called anywhere except from other methods that are called as part of coder inference; and this call site is one of the violations of this rule.

The fix to this place is to fix spark StreamingTransformTranslator.readUnbounded() to, again, get the coder from the output of the transform (from EvaluationContext) instead of from the source itself.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/StateSpecFunctions.java [hunk 17]
```diff
@@ -216,7 +217,7 @@
                 readDurationMillis,
                 metricsContainers));
 
-        } catch (IOException e) {
+        } catch (IOException | CannotProvideCoderException e) {
```
Comment (jkff) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/StateSpecFunctions.java]: Same comment here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java [hunk 18]
```diff
@@ -38,6 +38,7 @@
 import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
 import javax.annotation.Nullable;
+import org.apache.beam.sdk.coders.CannotProvideCoderException;
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java]: Source.getDefaultOutputCoder is also called from a bunch of other methods in SourceTestUtils. Those methods need to have new overloads where the user provides a Coder explicitly, for the same reason as above.

**Comment by jkff** [comment]
Actually solving this stuff is pretty complicated. For speed and because I need this PR in another PR, I'm going to finish this up myself and merge. Hope that's ok.

**Comment by jkff** [comment]
The changes were big enough that I moved them into a separate PR https://github.com/apache/beam/pull/3634 . I believe the current PR should be closed.

**Comment by lgajowy** [comment]
@jkff ok. Ping me if there is anything I can help you with this issue (or maybe some other issues). Good luck!

**Comment by jkff** [comment]
@lgajowy Would you mind closing the current PR?

**Comment by lgajowy** [comment]
@jkff thanks for reminding me. 👍
